### PR TITLE
tp: add minimal flatbuffer reader and writer

### PR DESCRIPTION
**Pull request by LalitMaganti**
Trace processor needs to emit and parse small flatbuffer-encoded
metadata (Arrow IPC framing) without taking a dependency on the
flatbuffers library, which would have to be vendored for every
embedder including the Wasm and Android builds.

The writer builds buffers back-to-front so offsets are naturally
forward-pointing. It tracks the maximum requested alignment and pads
the final size to a multiple of it: element positions are aligned
relative to the buffer end, so an aligned total size is what makes
them aligned relative to the start, which is what strict flatbuffers
verifiers check. String length prefixes are padded the same way.

The reader is designed for untrusted input: every table, vtable,
string and vector access is bounds-checked in 64-bit offset arithmetic
and malformed or absent data uniformly reads back as defaults.

### CODE REVIEW

File: src/trace_processor/util/flatbuffer_writer.cc [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  }
+  uint32_t num_slots = fields_.empty() ? 0 : max_index + 1;
+  uint32_t vtable_size = 4 + num_slots * 2;
+  uint32_t table_size = table_off - table_body_start_;
+
+  // Each vtable entry is the byte offset from the table start (soffset
+  // position) to the field data. 0 means "not present".
+  std::vector<uint16_t> vt_entries(num_slots, 0);
+  for (const auto& f : fields_) {
+    vt_entries[f.index] = static_cast<uint16_t>(table_off - f.off_from_end);
```
Comment (safayat-google) [inline src/trace_processor/util/flatbuffer_writer.cc:189]: nits, log some warning when table_off - f.off_from_end) > uint16_t max?

File: src/trace_processor/util/flatbuffer_reader.h [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+  // Reads a sub-table field (offset -> child table). Invalid reader if
+  // absent or malformed.
+  FlatBufferReader Table(uint32_t field_index) const;
+
+  // Reads a vector of inline scalars or packed structs of type T. Empty view
+  // if absent, malformed or extending past the end of the buffer.
+  template <typename T>
+  FlatBufferScalarVec<T> VecScalar(uint32_t field_index) const {
+    uint32_t count = 0;
```
Comment (safayat-google) [inline src/trace_processor/util/flatbuffer_reader.h:158]: nits, should we also check for out of bound similar to `FlatBufferTableVec`?

File: src/trace_processor/util/flatbuffer_reader_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+  auto reader =
+      FlatBufferReader::GetRoot(buf.data(), static_cast<uint32_t>(buf.size()));
+  ASSERT_TRUE(reader.has_value());
+  EXPECT_EQ(reader->String(0), "");
+  EXPECT_FALSE(reader->Table(0));
+  EXPECT_EQ(reader->VecScalar<int32_t>(0).size(), 0u);
+  EXPECT_EQ(reader->VecTable(0).size(), 0u);
+  EXPECT_EQ(reader->VecString(0).size(), 0u);
+}
+
```
Comment (safayat-google) [inline src/trace_processor/util/flatbuffer_reader_unittest.cc:452]: nits, should we add more tests to cover index access for out size or maximum boundary for major supported types?

**Review by safayat-google — APPROVED** [review]

File: src/trace_processor/util/flatbuffer_writer.h [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+  // Prepends a single scalar value.
+  template <typename T>
+  void Prepend(T val) {
+    GrowIfNeeded(sizeof(T));
+    head_ -= sizeof(T);
+    memcpy(&buf_[head_], &val, sizeof(T));
+  }
+
+  uint32_t GetSize() const {
```
Comment (sashwinbalaji) [inline src/trace_processor/util/flatbuffer_writer.h]: driveby comment: this is exactly same as Size() func above so can be removed.

File: src/trace_processor/util/flatbuffer_writer.cc (hunk 1 above)
Comment (LalitMaganti) [reply src/trace_processor/util/flatbuffer_writer.cc:189]: Done - EndTable now logs an error when the table size overflows the 16-bit vtable encoding. That single check also bounds every vtable entry, since each entry is smaller than table_size.

File: src/trace_processor/util/flatbuffer_reader.h (hunk 2 above)
Comment (LalitMaganti) [reply src/trace_processor/util/flatbuffer_reader.h:158]: Done - FlatBufferScalarVec::operator[] now returns a default-constructed value for out-of-range indices, matching FlatBufferTableVec/FlatBufferStringVec.

File: src/trace_processor/util/flatbuffer_reader_unittest.cc (hunk 3 above)
Comment (LalitMaganti) [reply src/trace_processor/util/flatbuffer_reader_unittest.cc:452]: Done - added VecIndexOutOfBoundsReturnsDefault covering the last valid index, one-past-the-end and extreme indices for scalar (int64/double/uint8), string and table vectors.

File: src/trace_processor/util/flatbuffer_writer.h (hunk 4 above)
Comment (LalitMaganti) [reply src/trace_processor/util/flatbuffer_writer.h]: Done - removed GetSize(); the writer uses size() internally now.
